packages/editor/src/components/provider/disable-non-page-content-blocks.js

```diff
@@ -32,38 +36,85 @@ export default function DisableNonPageContentBlocks() {
 
 	const registry = useRegistry();
 
+	// The code here is split into multiple `useEffects` calls.
+	// This is done to avoid setting/unsetting block editing modes multiple times unnecessarily.
+	//
+	// For example, the block editing mode of the root block (clientId: '') only
+	// needs to be set once, not when `contentOnlyIds` or `disabledIds` change.
+	//
+	// It's also unlikely that these different types of blocks are being inserted
+	// or removed at the same time, so using different effects reflects that.
+	useEffect( () => {
+		const { setBlockEditingMode, unsetBlockEditingMode } =
+			registry.dispatch( blockEditorStore );
+
+		setBlockEditingMode( '', 'disabled' );
+
+		return () => {
+			unsetBlockEditingMode( '' );
+		};
+	}, [ registry ] );
+
 	useEffect( () => {
 		const { setBlockEditingMode, unsetBlockEditingMode } =
 			registry.dispatch( blockEditorStore );
 
 		registry.batch( () => {
-			setBlockEditingMode( '', 'disabled' );
 			for ( const clientId of contentOnlyIds ) {
 				setBlockEditingMode( clientId, 'contentOnly' );
 			}
-			for ( const clientId of templateParts ) {
-				setBlockEditingMode( clientId, 'contentOnly' );
-			}
-			for ( const clientId of disabledIds ) {
-				setBlockEditingMode( clientId, 'disabled' );
-			}
 		} );
 
 		return () => {
 			registry.batch( () => {
-				unsetBlockEditingMode( '' );
 				for ( const clientId of contentOnlyIds ) {
 					unsetBlockEditingMode( clientId );
 				}
+			} );
+		};
+	}, [ contentOnlyIds, registry ] );
+
+	useEffect( () => {
+		const { setBlockEditingMode, unsetBlockEditingMode } =
+			registry.dispatch( blockEditorStore );
+
+		registry.batch( () => {
+			if ( ! isNavigationMode ) {
 				for ( const clientId of templateParts ) {
-					unsetBlockEditingMode( clientId );
+					setBlockEditingMode( clientId, 'contentOnly' );
 				}
+			}
+		} );
+
+		return () => {
+			registry.batch( () => {
+				if ( ! isNavigationMode ) {
+					for ( const clientId of templateParts ) {
+						unsetBlockEditingMode( clientId );
+					}
+				}
+			} );
+		};
+	}, [ templateParts, isNavigationMode, registry ] );
+
+	useEffect( () => {
+		const { setBlockEditingMode, unsetBlockEditingMode } =
+			registry.dispatch( blockEditorStore );
+
+		registry.batch( () => {
+			for ( const clientId of disabledIds ) {
+				setBlockEditingMode( clientId, 'disabled' );
+			}
+		} );
+
+		return () => {
+			registry.batch( () => {
 				for ( const clientId of disabledIds ) {
 					unsetBlockEditingMode( clientId );
 				}
 			} );
 		};
-	}, [ templateParts, contentOnlyIds, disabledIds, registry ] );
+	}, [ disabledIds, registry ] );
 
 	return null;
 }
```

Comment:
Yeah, it is confusing! There's already a list like this - https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/provider/use-post-content-blocks.js

I think your idea works for most of the blocks, apart from the template part where the block itself is enabled but inner blocks are disabled. This is buggy anyway, as disabling the inner blocks like that doesn't prevent insertions, so at the moment you can still drag things into the template part block when editing a page. 😓 

I personally think it'd be best to remove the way template parts are selectable in this mode. There's a parallel discussion on it here - https://github.com/WordPress/gutenberg/issues/67621#issuecomment-2522509917.

Reply:
Another thing is that in the long run, I think it'd be interesting to determine these blocks via their bindings.

E.g. if the user is editing a `page` with the id `12`, then any blocks with bindings to that entity become editable.

That's a bit far away, but it makes me thing that rather than `editableBlockTypes` we should instead pass `editableClientIds` as the editor setting.

Blocks with a post meta binding could already be made to work like this if they're in the template.

Reply:
I think all these exceptions we're introducing to please some specific UX affordance are really not helping. We keep constructing and deconstructing. I'm not sure how to solve it personally but I think we should consider scaling back UX wise maybe and trying to systemize these changes.

It's really not clear to me why we're treating template parts differently than reusable blocks. We should (at least in our minds) consider these two things the same.

Reply:
> Ok, I'll look into the template part / synced pattern disparity in a separate PR

#68042 is the PR.

Actually it does highlight something. Giving Site Title/Logo the `role: content` attribute isn't enough. In a pattern a paragraph would have this role, but it still isn't editable unless it is marked as a pattern override.

We likely need some additional mechanic to make this work.